src/sequential/sequential.hpp

```diff
@@ -0,0 +1,198 @@
+  // we can fit this bucket into 4 words, which is more amenable to the hardware.
+  // A value of SIZE_MAX would be attrocious for performance anyways.
+  size_t offset;
+
+  SequentialBucket(KeyType key, ValueType value, HashCodeType hashcode, size_t offset) {
+    LOG_TRACE("Enter");
+    this->key = key;
+    this->value = value;
+    this->hashcode = hashcode;
+    this->offset = offset;
+  }
+
+  SequentialBucket() {
+    LOG_TRACE("Enter");
+    this->key = 0;
+    this->value = 0;
+    this->hashcode = 0;
+    this->offset = SIZE_MAX;
+  }
+
+  bool is_empty();
+
+  void set(const KeyType key,
+           const ValueType value,
+           const HashCodeType hashcode,
+           const size_t offset);
+
+  void swap(SequentialBucket &other);
+
+  bool equal_key(const KeyType key, const HashCodeType hashcode);
+
+  /// @brief  Pretty print bucket
+  ///
+  /// This prints in the format (using Python's f-string syntax):
+  /// f"({hashcode}=>{home}+{offset}) {key}: {value}"
+  void print(const size_t size);
+};
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+/// Source: https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key
+HashCodeType
+hash(const KeyType key);
+
+size_t
+get_home(const HashCodeType hashcode, const size_t size);
+
+/// @brief  Get real bucket index.
+size_t
+get_probe(const size_t home, const size_t offset, const size_t size);
+
+/// @brief  Get offset from home or where it would be if not found.
+///         Return SIZE_MAX if no hole is found.
+std::pair<SearchStatus, size_t>
+get_wouldbe_offset(      std::vector<SequentialBucket> &tmp_buckets,
+                   const KeyType key,
+                   const HashCodeType hashcode,
+                   const size_t home);
+
+/// @brief  Insert but assume no resize is necessary.
+int
+insert_without_resize(      std::vector<SequentialBucket> &tmp_buckets,
+                          const KeyType key,
+                          const ValueType value,
+                          const HashCodeType hashcode);
+
+////////////////////////////////////////////////////////////////////////////////
+/// HASH TABLE CLASS
+////////////////////////////////////////////////////////////////////////////////
+
+class SequentialRobinHoodHashTable {
+private:
+  std::vector<SequentialBucket> buckets;
+  size_t length;
+  size_t size;
+
+
+  int resize(size_t new_size) {
+    LOG_TRACE("Enter");
+    std::vector<SequentialBucket> tmp_bkts;
+    tmp_bkts.resize(new_size);
+    assert(new_size >= length && "not enough room in new array!");
+    for (auto &bkt : this->buckets) {
+      if (!bkt.is_empty()) {
+        insert_without_resize(tmp_bkts, bkt.key, bkt.value, bkt.hashcode);
+      }
+    }
+    this->buckets = tmp_bkts;
```

Comment:
```suggestion
    this->buckets = std::move(tmp_bkts);
```
A plain assignment would do another heap allocation.

Reply:
I'll need to ask you what's going on here. I'm not super familiar with the move operator.

Reply:
Direct assignment of `std::vector` invokes the copy constructor, which does a deep copy in order to conform to C/C++ value semantics. If this weren't the case, we would get a shallow copy instead, and modifying the data of one value would reflect on the other, which most C/C++ programmers would find highly unexpected. Consider:
```c++
class String {
  char *data;
  ...
};

String a = "hello, world!";
const String b = a; // Looks like a value copy, but is actually a shallow copy, which implies b.data = a.data.

a[0] = 'Y';
assert(b[0] == 'h'); // Assertion failure.
```
So C++ introduces the copy constructor to let programmers define what should actually happen when a copy occurs:
```c++
class String {
  char *data;
  size_t length;

  // Copy constructor.
  String(const String &other) {
    size_t length = other.length;
    char *data = new char[length];
    std::memcpy(data, other.data, length);

    this->data = data;
    this->length = length;
  }

  // Technically we also need a copy assignment operator for the above example but you get the point.
};
```
But what if we don't actually want to duplicate the data, especially since doing so would be costly? I.e., we just want to assign it to another variable?

That's why C++11 introduced the concept of moves (apparently there were ways around this before C++11 but that's ancient history). Moving is closely related to swapping. For the purpose of understanding the change I proposed above, you can see it as the following:
```c++
  std::swap(this->buckets, tmp_buckets);
  // this->buckets now points to the data of tmp_buckets and vice versa.
  ...
} // End of function - tmp_buckets, which has the original this->buckets data, has its destructor invoked.
```

Reply:
I'll just use std::move. Whatever.